Internal error when identifier in except clause shadows imported module name

<!--
  Use this form only if mypy reports an "INTERNAL ERROR" and/or gives a traceback.
  Please include the traceback and all other messages below (use `mypy --show-traceback`).
-->

**Crash Report**

Mypy crashes while analysing a `try` statement in which one of the `except` clauses uses the same identifier as one of the imported modules.

**Traceback**

```
bug.py:3: error: INTERNAL ERROR -- Please try using mypy master on Github:
https://mypy.rtfd.io/en/latest/common_issues.html#using-a-development-mypy-build
Please report a bug at https://github.com/python/mypy/issues
version: 0.790
Traceback (most recent call last):
  File "C:\Users\User\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "C:\Users\User\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "mypy\checker.py", line 401, in accept
  File "mypy\nodes.py", line 1235, in accept
  File "mypy\checker.py", line 3350, in visit_try_stmt
  File "mypy\checker.py", line 3403, in visit_try_without_finally
  File "mypy\checker.py", line 2053, in check_assignment
  File "mypy\checker.py", line 2799, in check_lvalue
TypeError: mypy.nodes.Var object expected; got mypy.nodes.MypyFile
bug.py:3: : note: use --pdb to drop into pdb
```

**To Reproduce**

bug.py contents:

```python
import re

try:
    pass
except Exception as re:
    pass
```

Command: `mypy bug.py`

**Your Environment**

<!-- Include as many relevant details about the environment you experienced the bug in -->

- Mypy version used: 0.790
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.6.8
- Operating system and version: Windows 10 2004

<!--
You can freely edit this text, please remove all the lines
you believe are unnecessary.
-->

--- a/mypy/checker.py
+++ b/mypy/checker.py
@@ -2796,7 +2796,9 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
         index_lvalue = None
         inferred = None
 
-        if self.is_definition(lvalue):
+        if self.is_definition(lvalue) and (
+            not isinstance(lvalue, NameExpr) or isinstance(lvalue.node, Var)
+        ):
             if isinstance(lvalue, NameExpr):
                 inferred = cast(Var, lvalue.node)
                 assert isinstance(inferred, Var)
@@ -3420,7 +3422,8 @@ class TypeChecker(NodeVisitor[None], CheckerPluginInterface):
                                 source = ('(exception variable "{}", which we do not '
                                           'accept outside except: blocks even in '
                                           'python 2)'.format(var.name))
-                            cast(Var, var.node).type = DeletedType(source=source)
+                            if isinstance(var.node, Var):
+                                var.node.type = DeletedType(source=source)
                             self.binder.cleanse(var)
             if s.else_body:
                 self.accept(s.else_body)

--- a/test-data/unit/check-redefine.test
+++ b/test-data/unit/check-redefine.test
@@ -474,3 +474,12 @@ with A() as x:
     reveal_type(x) # N: Revealed type is "builtins.int"
 with B() as x:
     x = 0 # E: Incompatible types in assignment (expression has type "int", variable has type "str")
+
+
+[case testRedefineModuleAsException]
+import typing
+try:
+    pass
+except Exception as typing:
+    pass
+[builtins fixtures/exception.pyi]

--- a/test-data/unit/check-unreachable-code.test
+++ b/test-data/unit/check-unreachable-code.test
@@ -1305,3 +1305,19 @@ async def f_malformed_2() -> int:
 
 [typing fixtures/typing-full.pyi]
 [builtins fixtures/tuple.pyi]
+
+
+[case testConditionalTypeVarException]
+# every part of this test case was necessary to trigger the crash
+import sys
+from typing import TypeVar
+
+T = TypeVar("T", int, str)
+
+def f(t: T) -> None:
+    if sys.platform == "lol":
+        try:
+            pass
+        except BaseException as e:
+            pass
+[builtins fixtures/dict.pyi]

